### ISSUE TITLE
Variants of ApplyIf that use Result type instead of Bool

### ISSUE THREAD

**Issue by vadym-kl** [microsoft/QuantumLibraries#142]
**Is your feature request related to a problem? Please describe.**
Resource estimation of code involving measurements is a challenge for `ResourceEstimator` Q# simulator. Currently the only way to help `ResourceEstimator` to deal with measurement is to use `AssertProb` to provide measurement outcome probability. There are resource estimation scenarios where this is not sufficient.

**Describe the solution you'd like**
Adding combinators `ApplyIfOne`, `ApplyIfZero`, `ApplyIfElse` to `Microsoft.Quantum.Canon` will help to express patterns of code that `ResourceEstimator` has potential to handle. 
Suggested signatures are the following ( +usual `A, CA, C` versions) : 

```qsharp
operation ApplyIfOne<'T>( 
    measurementResult : Result, 
    (onResultOneOp : ('T => Unit), oneArg : 'T) 
) : Unit

operation ApplyIfZero<'T>( 
    measurementResult : Result,
    (onResultZeroOp : ('T => Unit), zeroArg : 'T) 
) : Unit

operation ApplyIfElse<'T,'U>(
    measurementResult : Result,
    (onResultZeroOp : ('T => Unit), zeroArg : 'T) ,
    (onResultOneOp : ('U => Unit), oneArg : 'U) 
) : Unit

// Example of calls: 
ApplyIfOne(res, (CX,(q1, q2)) ); 
// note how expression (CX, (q1, q2)) is reminiscent to usual call CX(q1,q2)
ApplyIfElse(res, (CX,(q1, q2)), (X,q3) );
```

**Describe alternatives you've considered**
One alternative is to replace type of `onResultZeroOp` and others by `Unit => Unit` and use 
```
operation HoldOperation<'T> ( op : ('T => Unit), arg : 'T, dummy : Unit ) : Unit {
  op(arg);
}
```
to reduce general case to `Unit => Unit`, but it seems to produce cumbersome code and has potential run-time overhead. 

**Additional context**
This is a prerequisite for potential improvements to `QCTraceSimulator` and `ResourceEstimator`. 
Feedback from Q# community and from @cgranade in particular about the choice of names and signatures is very welcome.

**Comment by vadym-kl** [comment]
Related PR #143

**Comment by cgranade** [comment]
Resolved with #143.
